File: Payload_Type/merlin/mythic/agent_functions/runas.py

from merlin import MerlinJob
from mythic_payloadtype_container.MythicCommandBase import *
from mythic_payloadtype_container.MythicRPC import *
import json

# Set to enable debug output to Mythic
debug = False


class RunAsArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="user",
                cli_name="user",
                display_name="Username",
                type=ParameterType.String,
                description="Domain and username to make a token for (e.g. ACME\\RASTLEY",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=1,
                        required=True,
                    ),
                ],
            ),
            CommandParameter(
                name="pass",
                cli_name="pass",
                display_name="Password",
                type=ParameterType.String,
                description="The account's password",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=2,
                        required=True,
                    ),
                ],
            ),
            CommandParameter(
                name="executable",
                cli_name="executable",
                display_name="Executable",
                type=ParameterType.String,
                description="The executable program to start",
                value="whoami",
                parameter_group_info=[ParameterGroupInfo(
                    group_name="Default",
                    ui_position=3,
                    required=True,
                )],
            ),
            CommandParameter(
                name="arguments",
                cli_name="args",
                display_name="Arguments",
                type=ParameterType.String,
                description="Arguments to start the executable with",
                parameter_group_info=[ParameterGroupInfo(
                    group_name="Default",
                    ui_position=4,
                    required=False,
                )],
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            if self.command_line[0] == '{':
                self.load_args_from_json_string(self.command_line)
            else:
                args = str.split(self.command_line)
                if len(args) < 3:
                    raise Exception("Expected 3 or more arguments")

                self.add_arg("user", args[0])
                self.add_arg("pass", args[1])
                self.add_arg("executable", args[2])
                if len(args) > 3:
                    self.add_arg("arguments", " ".join(args[3:]))


class RunAsCommand(CommandBase):
    cmd = "runas"
    needs_admin = False
    help_cmd = "runas"
    description = "Run the provided program as the user for the provided credentials"
    version = 1
    author = "@Ne0nd0g"
    argument_class = RunAsArguments
    attackmapping = ["T1106"]

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        if debug:
            await MythicRPC().execute(function_name="create_output", task_id=task.id, output=f'\n[DEBUG]Input task:{task}')

        # RunAs Arguments <user>, <pass>, <executable>, [<args>]
        args = [task.args.get_arg("user"), task.args.get_arg("pass"), task.args.get_arg("executable")]

        if task.args.get_arg("arguments"):
            task.display_params = f'User: {task.args.get_arg("user")} Pass: {task.args.get_arg("pass")} ' \
                                  f'Program: {task.args.get_arg("executable")} {task.args.get_arg("arguments")}'
            arguments = task.args.get_arg("arguments").split()
            if len(arguments) == 1:
                args.append(arguments[0])
            elif len(arguments) > 1:
                for arg in arguments:
                    args.append(arg)
        else:
            task.display_params = f'User: {task.args.get_arg("user")} Pass: {task.args.get_arg("pass")} ' \
                                  f'Program: {task.args.get_arg("executable")}'

        # Merlin jobs.Command message type
        command = {
            "command": self.cmd,
            "args": args,
        }

        task.args.remove_arg("user")
        task.args.remove_arg("pass")
        task.args.remove_arg("executable")
        task.args.remove_arg("arguments")
        task.args.add_arg("type", MerlinJob.MODULE, ParameterType.Number)
        task.args.add_arg("payload", json.dumps(command), ParameterType.String)

        if debug:
            await MythicRPC().execute("create_output", task_id=task.id, output=f'[DEBUG]Returned task:\r\n{task}\r\n')

        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/merlin/mythic/container/commands/runas.go
/*
Merlin is a post-exploitation command and control framework.

This file is part of Merlin.
Copyright (C) 2023  Russel Van Tuyl

Merlin is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as
published by the Free Software Foundation, either version 3 of the License, or any later version.

Merlin is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with Merlin.  If not, see <http://www.gnu.org/licenses/>.
*/

package commands

import (
	// Standard
	"fmt"
	"strings"

	// Mythic
	structs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"

	// Merlin
	"github.com/Ne0nd0g/merlin/pkg/jobs"
)

// runas creates and return a Mythic Command structure that is registered with the Mythic server
func runas() structs.Command {
	attr := structs.CommandAttribute{
		SupportedOS: []string{structs.SUPPORTED_OS_WINDOWS},
	}

	executable := structs.CommandParameter{
		Name:                                    "executable",
		ModalDisplayName:                        "Executable",
		CLIName:                                 "executable",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The executable, or program, to run",
		Choices:                                 nil,
		DefaultValue:                            nil,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       0,
				AdditionalInformation: nil,
			},
		},
	}

	args := structs.CommandParameter{
		Name:                                    "arguments",
		ModalDisplayName:                        "Arguments",
		CLIName:                                 "args",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "Arguments to start the executable with",
		Choices:                                 nil,
		DefaultValue:                            "",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
		},
	}

	user := structs.CommandParameter{
		Name:                                    "user",
		ModalDisplayName:                        "Username",
		CLIName:                                 "user",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "Domain and username to make a token for (e.g. ACME\\\\RASTLEY)",
		Choices:                                 nil,
		DefaultValue:                            nil,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
		},
	}

	pass := structs.CommandParameter{
		Name:                                    "password",
		ModalDisplayName:                        "Password",
		CLIName:                                 "pass",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The user account's password",
		Choices:                                 nil,
		DefaultValue:                            nil,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
		},
	}

	params := []structs.CommandParameter{executable, args, user, pass}
	command := structs.Command{
		Name:                           "runas",
		NeedsAdminPermissions:          false,
		HelpString:                     "runas <args> <username> <password>",
		Description:                    "Run the provided program as the user for the provided credentials",
		Version:                        0,
		SupportedUIFeatures:            nil,
		Author:                         "@Ne0nd0g",
		MitreAttackMappings:            []string{"T1106"},
		ScriptOnlyCommand:              false,
		CommandAttributes:              attr,
		CommandParameters:              params,
		AssociatedBrowserScript:        nil,
		TaskFunctionOPSECPre:           nil,
		TaskFunctionCreateTasking:      runasCreateTask,
		TaskFunctionProcessResponse:    nil,
		TaskFunctionOPSECPost:          nil,
		TaskFunctionParseArgString:     taskFunctionParseArgString,
		TaskFunctionParseArgDictionary: taskFunctionParseArgDictionary,
		TaskCompletionFunctions:        nil,
	}

	return command
}

// runasCreateTask takes a Mythic Task and converts into a Merlin Job that is encoded into JSON and subsequently sent to the Merlin Agent
func runasCreateTask(task *structs.PTTaskMessageAllData) (resp structs.PTTaskCreateTaskingMessageResponse) {
	pkg := "mythic/container/commands/runas/runasCreateTask()"
	resp.TaskID = task.Task.ID

	executable, err := task.Args.GetStringArg("executable")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'executable' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	args, err := task.Args.GetStringArg("arguments")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'arguments' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	user, err := task.Args.GetStringArg("user")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'user' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	pass, err := task.Args.GetStringArg("password")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'password' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	job := jobs.Command{
		Command: task.Task.CommandName,
		// RunAs Arguments <user>, <pass>, <executable>, [<args>]
		Args: []string{user, pass, executable},
	}

	if args != "" {
		job.Args = append(job.Args, strings.Split(args, " ")...)
	}

	mythicJob, err := ConvertMerlinJobToMythicTask(job, jobs.MODULE)
	if err != nil {
		err = fmt.Errorf("%s: there was an error converting the Merlin job to a Mythic job: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	task.Args.SetManualArgs(mythicJob)

	disp := fmt.Sprintf("Username: %s, Password: %s, %s %s", user, pass, executable, args)
	resp.DisplayParams = &disp
	resp.Success = true

	return
}
